File: slider/src/androidTest/java/com/adsamcik/slider/FloatSliderInstrumentationTest.java
package com.adsamcik.slider;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import junit.framework.Assert;

import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.*;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class FloatSliderInstrumentationTest {
	private final float DELTA = 0.00001f;

	@Test
	public void baseTest() throws Exception {
		// Context of the app under test.
		Context appContext = InstrumentationRegistry.getTargetContext();

		assertEquals("com.adsamcik.slider.test", appContext.getPackageName());

		final float MAX = 9;
		final float MIN = 1;
		final int PLACES = 1;
		final int POW = (int) (100 * Math.pow(10, PLACES));
		final float MAX_SLIDER = (MAX - MIN) * POW;

		FloatSlider slider = new FloatSlider(appContext);
		slider.setMinValue(MIN);
		slider.setMaxValue(MAX);
		slider.setStep(2f);
		slider.setValue(5f);
		assertEquals(MIN, slider.getMinValue(), DELTA);
		assertEquals(MAX, slider.getMaxValue(), DELTA);
		assertEquals(MAX_SLIDER, slider.getMax(), DELTA);
		assertEquals(4 * POW, slider.getProgress(), DELTA);
		assertEquals(5, slider.getValue(), DELTA);
	}

	@Test
	public void moveTest() throws Exception {
		// Context of the app under test.
		Context appContext = InstrumentationRegistry.getTargetContext();

		assertEquals("com.adsamcik.slider.test", appContext.getPackageName());

		FloatSlider slider = new FloatSlider(appContext);
		slider.setMinValue(0f);
		slider.setMaxValue(12f);
		slider.setStep(3f);
		for (int i = 0; i < 12; i++) {
			slider.setValue((float) i);
			assertEquals(Math.round(i / 3.0) * 3, slider.getValue(), DELTA);
		}
	}

	@Test
	public void negativeNumberTest() throws Exception {
		// Context of the app under test.
		Context appContext = InstrumentationRegistry.getTargetContext();

		assertEquals("com.adsamcik.slider.test", appContext.getPackageName());

		FloatSlider slider = new FloatSlider(appContext);
		slider.setMinValue(-15f);
		slider.setMaxValue(15f);
		slider.setStep(3f);
		for (int i = -15; i <= 15; i++) {
			slider.setValue((float) i);
			assertEquals(Math.round(i / 3f) * 3f, slider.getValue(), DELTA);
		}
	}

	@Test
	public void floatNumberTest() throws Exception {
		// Context of the app under test.
		Context appContext = InstrumentationRegistry.getTargetContext();

		assertEquals("com.adsamcik.slider.test", appContext.getPackageName());

		FloatSlider slider = new FloatSlider(appContext);
		slider.setMinValue(-1.5f);
		slider.setMaxValue(1.5f);
		slider.setStep(0.3f);
		for (float i = -1.5f; i <= 1.5f; i += 0.3f) {
			slider.setValue(i);
			assertEquals(Math.round(i / 0.3f) * 0.3f, slider.getValue(), DELTA);
		}
	}

	@Test
	public void valueTest() throws Exception {
		Context appContext = InstrumentationRegistry.getTargetContext();

		Slider<Float> slider = new FloatSlider(appContext);
		slider.setMinValue(-15f);
		slider.setMaxValue(15f);
		slider.setStep(3f);
		Float[] floats = new Float[]{0f, 3f, 4f, 20f, 35f};
		slider.setItems(floats);
		for (int i = 0; i < floats.length; i++) {
			slider.setValue(floats[i]);
			assertEquals(floats[i], slider.getValue());
		}

		for (int i = 0; i < floats.length; i++) {
			slider.setProgress(i);
			assertEquals(floats[i], slider.getValue());
		}

		EAssert.assertException(() -> slider.setMinValue(-15f), RuntimeException.class);
		EAssert.assertException(() -> slider.setMaxValue(15f), RuntimeException.class);
		EAssert.assertException(() -> slider.setStep(3f), RuntimeException.class);
	}

	@Test
	public void sharedPreferences() throws Exception {
		final Context appContext = InstrumentationRegistry.getTargetContext();
		final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(appContext);
		final String prefName = "TESTING PREFERENCE";

		FloatSlider slider = new FloatSlider(appContext);
		slider.setMinValue(-1.5f);
		slider.setMaxValue(1.5f);
		slider.setStep(0.3f);
		slider.setPreferences(preferences, prefName);
		slider.setValue(0.1f);

		Assert.assertEquals(slider.getValue(), preferences.getFloat(prefName, Float.MIN_VALUE));

		preferences.edit().remove(prefName).apply();
	}
}

File: slider/src/androidTest/java/com/adsamcik/slider/FloatSliderInstrumentationTest.kt
package com.adsamcik.slider

import android.content.Context
import android.widget.TextView
import androidx.test.core.app.ApplicationProvider
import com.adsamcik.slider.implementations.FloatSlider
import org.junit.Assert.assertEquals
import org.junit.Test
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */
class FloatSliderInstrumentationTest {
	companion object {
		private const val DELTA = 0.00001f
	}

	private val appContext = ApplicationProvider.getApplicationContext<Context>()!!

	@Suppress("LocalVariableName")
	@Test
	@Throws(Exception::class)
	fun baseTest() {
		val MAX = 9f
		val MIN = 1f
		val PLACES = 1

		val slider = FloatSlider(appContext)
		slider.minValue = MIN
		slider.maxValue = MAX
		slider.step = 2f
		slider.value = 5f
		assertEquals(MIN, slider.minValue, DELTA)
		assertEquals(MAX, slider.maxValue, DELTA)
		assertEquals(0.5f, slider.fluidPosition, DELTA)
		assertEquals(5f, slider.value, DELTA)
	}

	@Test
	@Throws(Exception::class)
	fun moveTest() {
		val slider = FloatSlider(appContext)
		slider.minValue = 0f
		slider.maxValue = 12f
		slider.step = 3f
		for (i in 0..11) {
			slider.value = i.toFloat()
			assertEquals(((i / 3.0).roundToInt() * 3).toFloat(), slider.value, DELTA)
		}
	}

	@Test
	@Throws(Exception::class)
	fun negativeNumberTest() {
		val slider = FloatSlider(appContext)
		slider.minValue = -15f
		slider.maxValue = 15f
		slider.step = 3f
		for (i in -15..15) {
			slider.value = i.toFloat()
			assertEquals((i / 3f).roundToInt() * 3f, slider.value, DELTA)
		}
	}

	@Test
	@Throws(Exception::class)
	fun floatNumberTest() {
		val slider = FloatSlider(appContext)
		slider.minValue = -1.5f
		slider.maxValue = 1.5f
		slider.step = 0.3f
		var i = -1.5f
		while (i <= 1.5f) {
			slider.value = i
			assertEquals((i / 0.3f).roundToInt() * 0.3f, slider.value, DELTA)
			i += 0.3f
		}
	}

	@Test
	@Throws(Exception::class)
	fun sharedPreferences() {
		val preferences = androidx.preference.PreferenceManager.getDefaultSharedPreferences(appContext)
		preferences.edit().clear().apply()

		val prefName = "TESTING PREFERENCE"

		val slider = FloatSlider(appContext)
		slider.minValue = -1.5f
		slider.maxValue = 1.5f
		slider.step = 0.3f
		slider.setPreferences(preferences, prefName, 1f)
		slider.value = 0.1f

		assertEquals(slider.value, preferences.getFloat(prefName, java.lang.Float.MIN_VALUE))

		preferences.edit().remove(prefName).apply()
	}
}
